### PR TITLE
Adding a resident-extern role

### LINKED ISSUE #347: Creating a resident extern status
It seems that the resident extern status is going to stay in the coming years, and it would be useful to track it. I quote @szentep:

_"Sziasztok! A mai igazgatói beszélgetésen világossá vált, hogy a BB státusz marad. Az utóbbi években jó megoldásnak bizonyult a helyek feltöltésére, hiszen ezek száma a felvételi végéig nem végleges, sőt, még szeptemberben is változik. Szerintem az Uránban is érdemes lenne bevezetni"_

As I remember, @machiato32 has said it would not be desirable as it would formalise something that should be abolished soon. But now, it looks like it's not going to be the case.

What do you think?

### PR DESCRIPTION
**Pull request by viktorcsimma**
Closes #347. It's largely done; there might be some bugs that I haven't noticed.

### CODE REVIEW

**Comment by viktorcsimma** [comment]
What should we do? I really think this would be useful; now that this is going to be the new status quo. @machiato32?

**Review by sszajbely — COMMENTED** [review]
The changes are alright. If this status really needs to stay, I guess I'll have to approve it. What do you think @kdmnk?

**Comment by sszajbely** [comment]
Also, please fix the merge conflicts @viktorcsimma

**Comment by viktorcsimma** [comment]
Something is not OK yet; it crashes at the economic committee page. I'll check it out soon.

**Comment by sszajbely** [comment]
Ah, sorry didn't test it, just looked at the code.

File: app/Http/Controllers/Secretariat/SemesterEvaluationController.php [hunk 1]
```diff
@@ -157,7 +157,7 @@ public function store(Request $request)
                         return back()->with('error', "A státusz megadása kötelező!")->with('section', $request->section);
                     }
                     $user->setStatusFor(Semester::next(), $request->next_status, $request->next_status_note);
-                    if ($request->has('resign_residency') && $user->isResident()) {
+                    if ($request->has('resign_residency') && $user->resides()) {
```
Comment (kdmnk) [inline app/Http/Controllers/Secretariat/SemesterEvaluationController.php:160]: resides() is a scope, which I think returns a query builder (test it in tinker). Check if these checks work as expected.

File: app/Models/ApplicationForm.php [hunk 2]
```diff
@@ -286,7 +289,7 @@ public function missingData(): array
             $missingData[] =  'Megjelölt kar';
         }
 
-        if (!$user->isResident() && !$user->isExtern()) {
+        if (!$user->resides() && !$user->isExtern()) {
```
Comment (kdmnk) [inline app/Models/ApplicationForm.php:292]: same

File: app/Http/Controllers/Secretariat/SemesterEvaluationController.php (hunk 1 above)
Comment (kdmnk) [inline app/Http/Controllers/Secretariat/SemesterEvaluationController.php:160]: resides() is a scope, which I think returns a query builder (test it in tinker). Check if these checks work as expected. In the tests it also seems fine, just double check it, I would not call scopes this way. I would rather make it a simple function

File: app/Models/ApplicationForm.php (hunk 2 above)
Comment (kdmnk) [inline app/Models/ApplicationForm.php:292]: same

**Review by kdmnk — COMMENTED** [review]
Can not we name this "LIMITED_RESIDENT" role internally? It is getting to be really confusing.
Alternatively, I was also thinking on making the roles time constrained. That way, we could have regular collegists, plus an additional resident role that is set to be valid for 1 year/indefinite.  We can also use that for guests, instead of the guest_until attribute and solve https://github.com/EotvosCollegium/mars/issues/291. (Also, the general collegist role is also valid only until the end of their studies by default which we can set this way).

@machiato32 what do you think, does putting effort into this would worth it?
(We have some time to work on this, the earliest this will be relevant is next summer)

**Comment by kdmnk** [comment]
> Can not we name this "LIMITED_RESIDENT" role internally? It is getting to be really confusing. Alternatively, I was also thinking on making the roles time constrained. That way, we could have regular collegists, plus an additional resident role that is set to be valid for 1 year/indefinite. We can also use that for guests, instead of the guest_until attribute and solve #291. (Also, the general collegist role is also valid only until the end of their studies by default which we can set this way).
> 
> @machiato32 what do you think, does putting effort into this would worth it? (We have some time to work on this, the earliest this will be relevant is next summer)

I'm starting to like this idea. The role_user table would have a created_at and valid_until field, and we would use a global scope to filter on those. This way, we don't delete roles just invalidate them, making a log out of the table.

**Comment by viktorcsimma** [comment]
> > Can not we name this "LIMITED_RESIDENT" role internally? It is getting to be really confusing. Alternatively, I was also thinking on making the roles time constrained. That way, we could have regular collegists, plus an additional resident role that is set to be valid for 1 year/indefinite. We can also use that for guests, instead of the guest_until attribute and solve #291. (Also, the general collegist role is also valid only until the end of their studies by default which we can set this way).
> > @machiato32 what do you think, does putting effort into this would worth it? (We have some time to work on this, the earliest this will be relevant is next summer)
> 
> I'm starting to like this idea. The role_user table would have a created_at and valid_until field, and we would use a global scope to filter on those. This way, we don't delete roles just invalidate them, making a log out of the table.

That sounds good:) Maybe we should open a new issue for that and freeze this until that gets solved.

**Comment by viktorcsimma** [comment]
What should we do with this one? Should we merge it and then reverse it when working on that issue, or abandon it altogether?

**Comment by kdmnk** [comment]
I'll close this for now.

**Pull request closed without merging**